contracts/domainRegistry.sol

```diff
@@ -0,0 +1,46 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.18;
+
+import "./domainsSetLibrary.sol";
+
+contract DomainRegistry {
+    uint256 public collateral = 0.001 ether;
+    using domainSet for domainSet.Set;
+
+    domainSet.Set internal domains;
+
+    event DomainRegistered(
+        string domainName,
+        address indexed owner,
+        uint256 deposit,
+        string operation
+    );
+
+    function registerDomain(string memory domainName) external payable {
+        require(msg.value == collateral, "Deposit must equal to collateral");
+        domains.add(domainName);
+        emit DomainRegistered(domainName, msg.sender, msg.value, "create");
+    }
+
+    function removeDomain(string memory domainName) external payable {
```

Comment:
There is still no restrictions to remove if the domain has child domains.   -1 point